Scripts/Runtime/Debug/UnityTraceListener.cs

```diff
@@ -0,0 +1,42 @@
+using System;
+using UnityEngine;
+
+namespace Anvil.Unity.Debugging
+{
+    /// <summary>
+    /// A custom <see cref="System.Diagnostics.TraceListener"> for Unity, to allow the use of <see cref="System.Diagnostics"> functions.
+    /// By default, all of the .NET Debug/Trace functions have no effect in Unity, without a listener manually forwarding calls.
+    /// 
+    /// The most common and important functions enabled by this listener are:
+    /// - <see cref="System.Diagnostics.Debug.WriteLine"> (forwarded to UnityEngine.Debug.Log)
+    /// - <see cref="System.Diagnostics.Debug.Assert"> (forwarded to a thrown exception)
+    /// 
+    /// Note: Both <see cref="System.Diagnostics.Debug"> and <see cref="System.Diagnostics.Trace"> functions use the same
+    /// listener, the difference is that Debug functions become no-ops in release builds, while Trace always works.
+    /// </summary>
+    internal class UnityTraceListener : System.Diagnostics.TraceListener
+    {
+        private static UnityTraceListener s_Instance;
+
+        [RuntimeInitializeOnLoadMethod(RuntimeInitializeLoadType.SubsystemRegistration)]
+        private static void Init()
+        {
+            if (s_Instance != null)
+            {
+                System.Diagnostics.Trace.Listeners.Remove(s_Instance);
+                s_Instance = null;
+            }
+
+            s_Instance = new UnityTraceListener();
+            System.Diagnostics.Trace.Listeners.Add(s_Instance);
+        }
+
+        private UnityTraceListener() { }
+
+        public override void Write(string message) { UnityEngine.Debug.Log(message); }
+        public override void WriteLine(string message) { UnityEngine.Debug.Log(message); }
```

Comment:
We should probably use Logger here. Something like:
`Log.GetStaticLogger(typeof(Test), "[ASSERT]")`

Although we caching the logger instance is likely our best bet.

If we use logger we can also include all of the "correct" context for the call site. Check out `UnityLogListener.ResolveCaller` for an example of this.

Reply:
I like the idea, but I'm not sure how to implement - where can I tell it to skip additional stack frames so the context falls on whoever called `System.Diagnostics.<something>` in the first place?

Would this replace the thrown exceptions too, or just the `Write`/`WriteLine` calls? I'm not sure where the `"[ASSERT]"` prefix comes in, and I assume `typeof(Test)` is a placeholder, but I don't know what context to actually use. If I try `Log.GetStaticLogger(typeof(UnityTraceListener)).Debug(message)` the output is the same as plain `Debug.Log(...)` after it gets picked up by Logger.

Reply:
> Where can I tell it to skip additional stack frames so the context falls on whoever called System.Diagnostics.<something> in the first place?

My thought is that you either copy the functionality of `UnityLogListener.ResolveCaller` and hard code the number of stack steps you'll be up from the original call site or break `UnityLogListener.ResolveCaller` out into its own utility method making the functionality available to all. I can't remember whether there are log listener details baked into that method but I can investigate late next week if needed.

You would then use the information from `ResolveCaller` to populate the optional arguments in `Logger.Debug` the way that we do in `UnityLogListener`

`UnityLogListener.SendToLogger` might be better to review to see how these pieces come together to build the context from stack frames. Again though, you'll have to either copy or break out this functionality. `UnityLogListener` itself isn't going to help you here and is really just the last line of defence in trying to pipe information through `Log`


> Would this replace the thrown exceptions too

No, I think the exception is preferred. It will should include the stack anyway.

> Assume typeof(Test) is a placeholder

oops yes `Test` should have been replaced before I commented. On second thought though you should be able to build a logger using the type from the original call site `UnityLogListener.ResolveContextFromMethod` does this.


> I'm not sure where the "[ASSERT]" prefix comes in

I figured it would be helpful to prefix the log messages with that but it's not necessary if the listener will be picking up non-assert messages.

Reply:
Thanks for the details - I'll give this a shot sometime before the PR is closed.

I've never seen `System.Diagnostics.Debug.WriteLine` used, I've always used `Console.WriteLine`, but a quick search tells me I've wrongly carried that practice over from C# console apps, and `Debug.WriteLine` is correct for debugging.

I don't see any `Console.WriteLine` uses in Anvil, but we do have a `ConsoleLogHandler` (presumably for console apps) - should there also be handlers for `System.Diagnostics.Debug` and `.Trace`? On the one hand, ideally all logs (within Anvil) use Anvil's Logger directly, but on the other hand Logger doesn't differentiate between `Debug` and `Trace` logging, and it doesn't make sense to use either (or both) for all logs in the app. 🤔

Reply:
Anything `*LogHandler` is for log outputting so `ConsoleLogHandler` is in fact for piping logger messages into `Console`. The `*LogListeners` are what capture logs made to other logging systems.

I suppose a `DebugLogHandler` is reasonable to have but I've never felt the need for one. Like you, I've always just carried over that `Console.WriteLine` was for debugging in console apps. 🤷 

Writing Log Listeners for Debug and Trace some day makes sense but isn't urgent until someone consumes a library that makes use of them and implementing the listener should be trivial with the hooks I'm seeing.

I'd say hold off on implementing any `Debug` or `Trace` log handlers for now.